packages/wiki-docs/gimp/GIMP-1-i18n.md

```diff
@@ -0,0 +1,119 @@
+* An `i18n:install` command is provided to copy `node_modules/@seasketch/geoprocessing/dist/base-project/i18n` for if you are upgrading to a new version of geoprocessing library and installing translations for the first time, or upgrading to the latest.
+* `createProject()` updates namespaces.json to a new project-specific namespace, `gp-{project_name}`.  The `gp` namespace is not present now because initial core translations are already installed and because in the most common use case we don't want projects publishing terms for the `gp` namespace.  That is left to be managed within the monorepo.  More on this later.
+
+```json
+{
+  "import": [
+    "gp-testo"
+  ],
+  "publish": [
+    "gp-testo"
+  ]
+}
+```
+
+## Client-side i18n load and switch
+
+* SSN will send a language code within `SeasketchReportingInitEvent` message.
+* It will pass the same code used in [SSN](https://github.com/seasketch/next/blob/master/packages/client/src/lang/supported.ts).
+* geoprocessing library will maintain its own version of that file in its i18n folder, and if SSN sends a code that's not supported it will gracefully fallback to the default language, with a console.warn
+* SSN will also send a new message to change language - `SeaSketchReportingChangeLanguageEvent` if the user switches.
+* App.tsx will read the language code from the postMessage event sent by SSN and store this in context.  If there is a change language event, then this code stored in context will be updated.  The language will default to `en` English and be updated with the initial language passed by SSN.
+* A `useLanguage()` hook will be provided that provides access to the current language stored in context and will trigger an update when the language changes.
+* each project client will load translations by wrapping a `Translator` react component, which creates an `i18n` instance (using i18nAsync.ts that lazy loads languages by default but i18nSync.ts is also available) and passes it in to `i18nextProvider`.
+* There is one i18n instance for each report client so that they are isolated and multiple can be loaded in storybook, or even an app at once.
+* The Translator will listen for language changes using the useLanguage hook.  If it changes, it will tell the i18n instance to change it's language, which will load the appropriate resource file and all translations in the UI will update.
+
+## Storybook
+
+* Translations just work in storybook.  Its webpack setup supports dynamic imports.
+* Since reports won't have a language switcher UI component, storybook should provide one. Just like for loading a sketch, it will add a default language to the mocked context provider, and change the language value when the language switcher.  This will update the useLanguage hook and trigger a re-render of everything.
+
+## For all report development use cases
+
+* the core gp translations will be installed on init or via `i18n:install` cli command and they will be ready to start.
```

Comment:
Does it make sense to run i18n:install as a post install package.json script so that projects are always up to date when upgrading? Is that command "safe" so that it won't blow away customized gp namespace translations if they exist?

Reply:
It is meant to be safe, haven't tested thoroughly.  Here's the command.  
```bash
"install:i18n": "mkdir -p src/i18n && cp -r node_modules/@seasketch/geoprocessing/dist/base-project/src/i18n src/i18n",
```

It's meant to overwrite existing files so in theory it will only copy in new gp.json files leaving the project namespace files intact.

Reply:
This has been improved.  base language translations are now bundled alongside project translations, which get extracted the first time on install.  They get merged into one set at runtime.

Here's the command to install/reinstall the base translations, copying them out of dist
```json
"translation:install": "mkdir -p src/i18n/baseLang && cp -r node_modules/@seasketch/geoprocessing/dist/base-project/src/i18n/baseLang src/i18n",
```